design/envoy-shutdown.md

```diff
@@ -0,0 +1,51 @@
+# Envoy Shutdown Probes
+
+Status: _Draft_
+
+This proposal describes how to add better support for identifying the status of Envoy connections during a restart or redeploy of Envoy.
+
+## Goals
+
+- Provide a mechanism to provide feedback on open connections & load inside an Envoy process
+- Allow for a Envoy fleet roll-out minimizing connection loss
+
+## Non Goals
+
+- Guarantee zero connection loss during a roll-out
+
+## Background
+
+The Envoy process, the data path component of Contour, at times needs to be re-deployed.
+This could be due to an upgrade, a change in configuration, or a node-failure forcing a redeployment.
+
+Contour currently implements a `preStop` hook in the container which signals to Envoy to begin draining connections.
+Once Envoy begins draining, the readiness probe on the pod triggers to fail, which in turn causes that instance of Envoy to stop accepting new connections.
+
+The main problem is that the `preStop` hook (which sends the `/healthcheck/fail` request) does not wait until Envoy has drained all connections, so when the container
+is restarted, users can receive errors.
+
+This design looks to add a new component to Contour which will allow for a way to understand if open connections exist in Envoy before sending a `SIGTERM`.
+
+## High-Level Design
+
+Implement a new sub-command of Contour named `envoyshutdown` which will handle sending the healthcheck fail request to Envoy and then begin polling the http/https listeners for active connections from 
+the `/stats` endpoint available on `localhost:9001`.
+
+## Detailed Design
+
+- Implement new subcommand in Contour called `envoyshutdown`
+- This command will expose an http endpoint over a specific port (e.g. `8090`) and path `/shutdown`. 
+- A new container will be added to the Envoy Daemonset pod which will run this new sub-command (e.g. `contour envoyhealthcheck`)
```

Comment:
Since this is a second pod with a long-running process in it, what precautions should it take to avoid preventing pod shutdown? i.e. if envoy exits we probable want the pod to exit

Reply:
We could add a liveliness probe in this new pod to check for its health. 

I think managing what to do if envoy would exit is out of scope for this proposal.

Reply:
My concern is that if envoy dies and the shutdown manager does not, the pod won't exit and will never be re-scheduled. Conversely, to shutdown envoy gracefully, we need the manager pod to always be available. so it probably needs a restart policy?